File: CLI/XlsxCommand.cs
using System;
using System.IO;
using System.Threading.Tasks;
using Services.Concrete.Excel;
using Services.Exceptions.Export;

namespace CLI
{
    internal class XlsxCommand : ExportCommand
    {
        public const string COMMAND_NAME = "xlsx";
        private bool _exportIntoSingleFile = false;
        private readonly ExcelService _excelService;

        public XlsxCommand() : base(COMMAND_NAME, @"Export demos into XLSX files")
        {
            _excelService = new ExcelService();
        }

        public override void PrintHelp()
        {
            Console.WriteLine(GetDescription());
            Console.WriteLine(@"");
            Console.WriteLine($@"Usage: {Program.ExeName} {COMMAND_NAME} demoPaths... [--output] [--source] [--single] [--steamid] [--force-analyze]");
            Console.WriteLine(@"");
            Console.WriteLine(@"Demos path can be either .dem files location or a directory. It can be relative or absolute.");
            Console.WriteLine(@"The --output argument specify the directory where output files will be saved.");
            Console.WriteLine($@"The --source argument force the analysis logic of the demo analyzer. Available values: [{string.Join(",", _availableSources)}]");
            Console.WriteLine(@"The --single argument generates a single XLSX file instead of one per demo.");
            Console.WriteLine(@"The --steamid argument makes data in some sheets focusing on the player with the given SteamID (works with --single only).");
            Console.WriteLine(@"The --force-analyze argument force demos analyzes (ignore cached data).");
            Console.WriteLine(@"");
            Console.WriteLine(@"Examples:");
            Console.WriteLine(@"");
            Console.WriteLine(@"Export a demo:");
            Console.WriteLine($@"    {Program.ExeName} {COMMAND_NAME} ""C:\Users\username\Desktop\demo.dem""");
            Console.WriteLine(@"");
            Console.WriteLine(@"Export multiple demos:");
            Console.WriteLine($@"    {Program.ExeName} {COMMAND_NAME} ""C:\Users\username\Desktop\demo.dem"" ""C:\Users\username\Desktop\demo2.dem""");
            Console.WriteLine(@"");
            Console.WriteLine(@"Export all demos in a directory using the ESL analyzer and save it in a custom directory:");
            Console.WriteLine($@"    {Program.ExeName} {COMMAND_NAME} ""C:\Users\username\Desktop\MyFolder"" --output ""C:\Users\username\Documents"" --source esl");
        }

        private new void ParseArgs(string[] args)
        {
            ParseArgs(args, new string[] { "--single" });
            for (int index = 0; index < args.Length; index++)
            {
                if (args[index] == "--single")
                {
                    _exportIntoSingleFile = true;
                    break;
                }
            }
        }

        public override async Task Run(string[] args)
        {
            ParseArgs(args);

            if (_demoPaths.Count == 0)
            {
                Console.WriteLine(@"No demo paths provided");
                return;
            }

            try
            {
                if (_exportIntoSingleFile)
                {
                    await ProcessSingleFileExport();
                }
                else
                {
                    await ProcessMultipleFilesExport();
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($@"Error while exporting demo: {ex.Message}");
                Environment.Exit(1);
            }
        }

        private async Task ProcessSingleFileExport()
        {
            if (string.IsNullOrEmpty(_outputFolderPath))
            {
                if (IsCurrentDirectoryWritable())
                {
                    _outputFolderPath = Directory.GetCurrentDirectory();
                }
                else
                {
                    _outputFolderPath = Environment.GetFolderPath(Environment.SpecialFolder.Desktop);
                }
            }
            string fileName = _outputFolderPath + Path.DirectorySeparatorChar + "export-" + DateTime.Now.ToString("yy-MM-dd-hh-mm-ss") + ".xlsx";
            MultiExportConfiguration configuration = new MultiExportConfiguration
            {
                FileName = fileName,
                DemoPaths = _demoPaths,
                FocusSteamId = _focusSteamId,
                Source = _source,
                ForceAnalyze = _forceAnalyze,
                OnProcessingDemo = (demoPath, demoNumber, totalDemoCount) => Console.WriteLine($@"Retrieving demo {demoNumber}/{totalDemoCount} {demoPath}"),
                OnDemoNotFound = demoPath => Console.WriteLine($@"The demo doesn't exists: {demoPath}"),
                OnInvalidDemo = demoPath => Console.WriteLine($@"Invalid demo {demoPath}"),
                OnAnalyzeStart = demoPath => Console.WriteLine($@"Analyzing demo {demoPath}"),
                OnAnalyzeError = demoPath => Console.WriteLine($@"Error while analyzing demo {demoPath}"),
                OnGeneratingXlsxFile = () => Console.WriteLine(@"Generating XLSX file..."),

            };
            await _excelService.GenerateXls(configuration);
            Console.WriteLine($@"XLSX file generated at {fileName}");
        }

        private async Task ProcessMultipleFilesExport()
        {
            int currentDemoNumber = 0;
            foreach (string demoPath in _demoPaths)
            {
                try
                {
                    Console.WriteLine($@"Retrieving demo {++currentDemoNumber}/{_demoPaths.Count} {demoPath}");
                    string outputFolderPath = BuildOutputFolderPathFromDemoPath(demoPath);
                    string demoName = Path.GetFileName(demoPath);
                    string fileName = outputFolderPath + Path.DirectorySeparatorChar + demoName + ".xlsx";
                    SingleExportConfiguration configuration = new SingleExportConfiguration
                    {
                        DemoPath = demoPath,
                        FileName = fileName,
                        Source = _source,
                        ForceAnalyze = _forceAnalyze,
                        OnAnalyzeStart = () => Console.WriteLine($@"Analyzing demo {demoPath}"),
                    };
                    await _excelService.GenerateXls(configuration);
                    Console.WriteLine($@"XLSX file generated at {fileName}");
                }
                catch (Exception ex)
                {
                    switch (ex)
                    {
                        case FileNotFoundException _:
                            Console.WriteLine($@"The demo doesn't exists: {demoPath}");
                            break;
                        case InvalidDemoException _:
                            Console.WriteLine($@"Invalid demo {demoPath}");
                            break;
                        case AnalyzeException _:
                            Console.WriteLine($@"Error while analyzing demo {demoPath}");
                            break;
                        default:
                            Console.WriteLine($@"An error occurred while exporting demo {demoPath}");
                            break;
                    }
                }
            }
        }
    }
}

File: src/cli/commands/xlsx-command.ts
import path from 'node:path';
import { SupportedDemoSources } from 'csdm/common/types/counter-strike';
import { MatchXlsxExport } from 'csdm/node/xlsx/match-xlsx-export';
import { MatchesXlsxExport } from 'csdm/node/xlsx/matches-xlsx-export';
import { SheetName } from 'csdm/node/xlsx/sheet-name';
import { ExportCommand } from './export-command';

export class XlsxCommand extends ExportCommand {
  public static Name = 'xlsx';
  private generateSingleFile = false;
  private sheets: Record<SheetName, boolean> = {
    [SheetName.General]: true,
    [SheetName.Players]: true,
    [SheetName.Rounds]: true,
    [SheetName.Kills]: true,
    [SheetName.Weapons]: true,
    [SheetName.Clutches]: true,
    [SheetName.Utility]: true,
    [SheetName.PlayersFlashbangMatrix]: true,
  };
  private singleFlag = '--single';
  private sheetsFlag = '--sheets';

  public getDescription() {
    return 'Export demos into XLSX files.';
  }

  public printHelp() {
    const sourceValues = SupportedDemoSources.join(',');
    console.log(this.getDescription());
    console.log('');
    console.log(
      `Usage: csdm ${XlsxCommand.Name} demoPaths... ${this.commonFlags.map((flag) => `[${flag}]`)} [${this.singleFlag}] [${this.sheetsFlag}]`,
    );
    console.log('');
    console.log('Demos path can be either a .dem files path or a directory. It can be relative or absolute.');
    console.log(`The ${this.outputFolderFlag} flag specify the directory where output files will be saved.`);
    console.log(
      `The ${this.sourceFlag} flag forces the analysis logic of the demo analyzer. Available values: [${sourceValues}]`,
    );
    console.log(`The ${this.singleFlag} flag generates a single XLSX file instead of one per demo.`);
    console.log(
      `The ${this.sheetsFlag} flag includes only sheets identified by its name. Available values: [${Object.values(
        SheetName,
      ).join(',')}]`,
    );
    console.log(`The ${this.forceAnalyzeFlag} flag forces demos analyzes even if they are already in the database.`);
    console.log('');
    console.log('Examples:');
    console.log('');
    console.log('Export 1 demo:');
    console.log(`    csdm ${XlsxCommand.Name} "C:\\Users\\username\\Desktop\\demo.dem"`);
    console.log('');
    console.log('Export multiple demos with only the "Rounds" and "Kills" sheets:');
    console.log(
      `    csdm ${XlsxCommand.Name} "C:\\Users\\username\\Desktop\\demo.dem" "C:\\Users\\username\\Desktop\\demo2.dem" ${this.sheetsFlag} rounds,kills`,
    );
    console.log('');
    console.log('Export all demos in a directory using the ESL analyzer and save it in a custom directory:');
    console.log(
      `    csdm ${XlsxCommand.Name} "C:\\Users\\username\\Desktop\\MyFolder" ${this.outputFolderFlag} "C:\\Users\\username\\Documents" ${this.sourceFlag} esl`,
    );
  }

  public async run() {
    await this.parseArgs();
    await super.run();

    if (this.generateSingleFile) {
      const checksums = this.demosToExport.map((demo) => demo.checksum);
      const outputFilePath = path.join(this.outputFolderPath, `export-${checksums.length}-matches.xlsx`);
      const xlsxExport = new MatchesXlsxExport({
        checksums,
        outputFilePath,
        sheets: this.sheets,
      });
      await xlsxExport.generate();
      console.log(`Exported ${checksums.length} matches in ${outputFilePath}`);
    } else {
      for (const { checksum, demoName } of this.demosToExport) {
        const outputFilePath = path.join(this.outputFolderPath, `${demoName}.xlsx`);
        const xlsxExport = new MatchXlsxExport({
          checksum,
          outputFilePath,
          sheets: this.sheets,
        });
        await xlsxExport.generate();
      }
      console.log(`Exported ${this.demosToExport.length} matches in folder ${this.outputFolderPath}`);
    }
  }

  protected async parseArgs() {
    await super.parseArgs();

    for (let index = 0; index < this.args.length; index++) {
      const arg = this.args[index];
      if (!this.isFlagArgument(arg)) {
        continue;
      }

      switch (arg) {
        case this.singleFlag:
          this.generateSingleFile = true;
          break;
        case this.sheetsFlag:
          if (this.args.length > index + 1) {
            index += 1;
            for (const sheetName in this.sheets) {
              this.sheets[sheetName as SheetName] = false;
            }

            const sheetNames = this.args[index].split(',');
            for (const sheetName of sheetNames) {
              if (!Object.values(SheetName).includes(sheetName as SheetName)) {
                console.log(`Invalid sheet name "${sheetName}"`);
                console.log(`Available values: ${Object.values(SheetName).join(',')}`);
                this.exitWithFailure();
              }

              this.sheets[sheetName as SheetName] = true;
            }
          } else {
            console.log(`Missing ${this.sheetsFlag} value`);
            this.exitWithFailure();
          }
          break;
        default:
          if (!super.isCommonFlag(arg)) {
            console.log(`Unknown flag: ${arg}`);
            this.exitWithFailure();
          }
      }
    }
  }
}
